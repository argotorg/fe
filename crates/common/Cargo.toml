[package]
name = "fe-common"
version = "0.12.0-alpha"
authors = ["The Fe Developers <snakecharmers@ethereum.org>"]
edition = "2021"
license = "Apache-2.0"
repository = "https://github.com/ethereum/fe"

[dependencies]
fe-library = {path = "../library", version = "^0.12.0-alpha"}
tiny-keccak = { version = "2.0", features = ["keccak"] }
hex = "0.4"
codespan-reporting = "0.11.1"
serde = { version = "1", features = ["derive"] }
ron = "0.5.1"
difference = "2.0"
num-traits = "0.2.14"
once_cell = "1.8.0"
include_dir = "0.6.0"
indexmap = "1.6.2"
smol_str = "0.1.21"
