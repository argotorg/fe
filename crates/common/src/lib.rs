pub mod config;
pub mod core;
pub mod diagnostics;
pub mod file;
pub mod indexmap;
pub mod ingot;
pub mod urlext;

use file::FileIndex;

#[salsa::db]
// Each database must implement InputDb explicitly with its own storage mechanism
pub trait InputDb: salsa::Database {
    fn file_index(&self) -> FileIndex;
}

#[doc(hidden)]
pub use paste::paste;

/// Macro for implementing the InputDb trait for a Salsa database struct
///
/// This assumes the database has a field named `index` of type `Option<FileIndex>`.
///
/// # Example
///
/// ```
/// #[derive(Clone)]
/// #[salsa::db]
/// pub struct MyDatabase {
///     storage: salsa::Storage<Self>,
///     index: Option<FileIndex>,
/// }
///
/// impl_input_db!(MyDatabase);
/// ```
#[macro_export]
macro_rules! impl_input_db {
    ($db_type:ty) => {
        #[salsa::db]
        impl $crate::InputDb for $db_type {
            fn file_index(&self) -> $crate::file::FileIndex {
                self.index.clone().expect("FileIndex not initialized")
            }
        }
    };
}

/// Macro for implementing Default for a Salsa database with FileIndex
///
/// This assumes the database has a field named `index` of type `Option<FileIndex>`,
/// and will initialize it properly.
///
/// # Example
///
/// ```
/// #[derive(Clone)]
/// #[salsa::db]
/// pub struct MyDatabase {
///     storage: salsa::Storage<Self>,
///     index: Option<FileIndex>,
/// }
///
/// impl_db_default!(MyDatabase);
/// ```
#[macro_export]
macro_rules! impl_db_default {
    ($db_type:ty) => {
        impl Default for $db_type
        where
            $db_type: $crate::core::HasBuiltinCore,
        {
            fn default() -> Self {
                let mut db = Self {
                    storage: salsa::Storage::default(),
                    index: None,
                };
                let index = $crate::file::FileIndex::default(&db);
                db.index = Some(index);
                $crate::core::HasBuiltinCore::initialize_builtin_core(&mut db);
                db
            }
        }
    };
}

/// Macro for creating a standard Salsa database with FileIndex support
///
/// This creates a struct with:
/// - A `storage` field for Salsa
/// - An `index` field for the FileIndex
/// - Default implementation that initializes the FileIndex
/// - Implementation of InputDb via impl_input_db!
///
/// # Example
///
/// ```
/// define_input_db!(MyDatabase);
/// ```
#[macro_export]
macro_rules! define_input_db {
    ($db_name:ident) => {
        #[derive(Clone)]
        #[salsa::db]
        pub struct $db_name {
            storage: salsa::Storage<Self>,
            index: Option<$crate::file::FileIndex>,
        }

        #[salsa::db]
        impl salsa::Database for $db_name {
            fn salsa_event(&self, _event: &dyn Fn() -> salsa::Event) {}
        }

        $crate::impl_input_db!($db_name);
        $crate::impl_db_default!($db_name);
    };
}

#[macro_export]
macro_rules! impl_db_traits {
    ($db_type:ty, $($trait_name:ident),+ $(,)?) => {
        #[salsa::db]
        impl salsa::Database for $db_type {
            fn salsa_event(&self, _event: &dyn Fn() -> salsa::Event) {}
        }
    };
}
