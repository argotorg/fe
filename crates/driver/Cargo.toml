[package]
name = "fe-driver"
version = "0.13.0-alpha"
authors = ["The Fe Developers <snakecharmers@ethereum.org>"]
edition = "2021"
license = "GPL-3.0-or-later"
repository = "https://github.com/ethereum/fe"

[features]
# Includes the solidity compiler for Yul -> bytecode compilation.
solc-backend = ["fe-yulc"]

[dependencies]
serde_json = "1.0"
fe-abi = {path = "../abi", version = "^0.13.0-alpha"}
fe-analyzer = {path = "../analyzer", version = "^0.13.0-alpha"}
fe-common = {path = "../common", version = "^0.13.0-alpha"}
fe-lowering = {path = "../lowering", version = "^0.13.0-alpha"}
fe-mir = {path = "../mir", version = "^0.13.0-alpha"}
fe-parser = {path = "../parser", version = "^0.13.0-alpha"}
fe-yulgen = {path = "../yulgen", version = "^0.13.0-alpha"}
fe-yulc = {path = "../yulc", version = "^0.13.0-alpha", features = ["solc-backend"], optional = true}
indexmap = "1.6.2"
vfs = "0.5.1"
smol_str = "0.1.21"
toml = "0.5.8"
