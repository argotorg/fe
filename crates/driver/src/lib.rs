use fe_analyzer::context::Analysis;
use fe_analyzer::namespace::items::{IngotId, ModuleId};
use fe_common::diagnostics::Diagnostic;
use fe_common::files::{FileStore, SourceFileId};
use fe_parser::ast::SmolStr;
use fe_yulgen::Db;
use indexmap::IndexMap;
#[cfg(feature = "solc-backend")]
use serde_json::Value;
use std::ops::Deref;

/// The artifacts of a compiled module.
pub struct CompiledModule {
    pub src_ast: String,
    pub lowered_ast: String,
    pub contracts: IndexMap<String, CompiledContract>,
}

/// The artifacts of a compiled contract.
pub struct CompiledContract {
    pub json_abi: String,
    pub yul: String,
    #[cfg(feature = "solc-backend")]
    pub bytecode: String,
}

#[derive(Debug)]
pub struct CompileError(pub Vec<Diagnostic>);

/// Compiles a single input file.
///
/// If `with_bytecode` is set to false, the compiler will skip the final Yul ->
/// Bytecode pass. This is useful when debugging invalid Yul code.
pub fn compile_module(
    files: &FileStore,
    file_id: SourceFileId,
    deps: &IndexMap<SmolStr, Vec<SourceFileId>>,
    _with_bytecode: bool,
    _optimize: bool,
) -> Result<CompiledModule, CompileError> {
    let mut errors = vec![];

    let db = Db::default();

    let Analysis {
        value: module_id,
        diagnostics: parser_diagnostics,
    } = ModuleId::try_new(&db, files, file_id, deps).map_err(CompileError)?;
    errors.extend(parser_diagnostics.deref().clone());

    match fe_analyzer::analyze_module(&db, module_id) {
        Ok(_) => {}
        Err(diagnostics) => {
            errors.extend(diagnostics.into_iter());
            return Err(CompileError(errors));
        }
    };

    if !errors.is_empty() {
        // There was a non-fatal parser error (eg missing parens in a fn def `fn foo: ...`)
        return Err(CompileError(errors));
    }

    // build abi
    let json_abis = fe_abi::build(&db, module_id).expect("failed to generate abi");

    // lower the AST
    let lowered_module_id = fe_lowering::lower_module(&db, module_id);
    let lowered_ast = format!("{:#?}", &lowered_module_id.ast(&db));

    fe_analyzer::analyze_module(&db, lowered_module_id).expect("failed to analyze lowered AST");

    // compile to yul
    let yul_contracts = fe_yulgen::compile(&db, lowered_module_id);

    // compile to bytecode if required
    #[cfg(feature = "solc-backend")]
    let bytecode_contracts = if _with_bytecode {
        match fe_yulc::compile(yul_contracts.clone(), _optimize) {
            Err(error) => {
                for error in serde_json::from_str::<Value>(&error.0)
                    .expect("unable to deserialize json output")["errors"]
                    .as_array()
                    .expect("errors not an array")
                {
                    eprintln!(
                        "Error: {}",
                        error["formattedMessage"]
                            .as_str()
                            .expect("error value not a string")
                            .replace("\\\n", "\n")
                    )
                }
                panic!("Yul compilation failed with the above errors")
            }
            Ok(contracts) => contracts,
        }
    } else {
        IndexMap::new()
    };

    // combine all of the named contract maps
    let contracts = json_abis
        .keys()
        .map(|name| {
            (
                name.to_owned(),
                CompiledContract {
                    json_abi: json_abis[name].to_owned(),
                    yul: yul_contracts[name].to_owned(),
                    #[cfg(feature = "solc-backend")]
                    bytecode: if _with_bytecode {
                        bytecode_contracts[name].to_owned()
                    } else {
                        "".to_string()
                    },
                },
            )
        })
        .collect::<IndexMap<_, _>>();

    Ok(CompiledModule {
        src_ast: format!("{:?}", module_id.ast(&db)),
        lowered_ast,
        contracts,
    })
}

/// Compiles a set of input files.
///
/// If `with_bytecode` is set to false, the compiler will skip the final Yul ->
/// Bytecode pass. This is useful when debugging invalid Yul code.
pub fn compile_ingot(
    name: &str,
    files: &FileStore,
    file_ids: &[SourceFileId],
    deps: &IndexMap<SmolStr, Vec<SourceFileId>>,
    _with_bytecode: bool,
    _optimize: bool,
) -> Result<CompiledModule, CompileError> {
    let mut errors = vec![];

    let db = Db::default();

    let Analysis {
        value: ingot_id,
        diagnostics: parser_diagnostics,
    } = IngotId::try_new(&db, files, name, file_ids, deps).map_err(CompileError)?;
    errors.extend(parser_diagnostics.deref().clone());

    match fe_analyzer::analyze_ingot(&db, ingot_id) {
        Ok(_) => {}
        Err(diagnostics) => {
            errors.extend(diagnostics.into_iter());
            return Err(CompileError(errors));
        }
    };

    if !errors.is_empty() {
        // There was a non-fatal parser error (eg missing parens in a fn def `fn foo: ...`)
        return Err(CompileError(errors));
    }

    let module_id = ingot_id.main_module(&db).expect("missing main module");

    // build abi
    let json_abis = fe_abi::build(&db, module_id).expect("failed to generate abi");
    let src_ast = format!("{:#?}", &module_id.ast(&db));

    let lowered_ingot_id = fe_lowering::lower_ingot(&db, ingot_id);

    fe_analyzer::analyze_ingot(&db, lowered_ingot_id).expect("failed to analyze lowered AST");

    let lowered_module_id = lowered_ingot_id
        .main_module(&db)
        .expect("missing main module");

    let lowered_ast = format!("{:#?}", &lowered_module_id.ast(&db));

    // compile to yul
    let yul_contracts = fe_yulgen::compile(&db, lowered_module_id);

    // compile to bytecode if required
    #[cfg(feature = "solc-backend")]
    let bytecode_contracts = if _with_bytecode {
        match fe_yulc::compile(yul_contracts.clone(), _optimize) {
            Err(error) => {
                for error in serde_json::from_str::<Value>(&error.0)
                    .expect("unable to deserialize json output")["errors"]
                    .as_array()
                    .expect("errors not an array")
                {
                    eprintln!(
                        "Error: {}",
                        error["formattedMessage"]
                            .as_str()
                            .expect("error value not a string")
                            .replace("\\\n", "\n")
                    )
                }
                panic!("Yul compilation failed with the above errors")
            }
            Ok(contracts) => contracts,
        }
    } else {
        IndexMap::new()
    };

    // combine all of the named contract maps
    let contracts = json_abis
        .keys()
        .map(|name| {
            (
                name.to_owned(),
                CompiledContract {
                    json_abi: json_abis[name].to_owned(),
                    yul: yul_contracts[name].to_owned(),
                    #[cfg(feature = "solc-backend")]
                    bytecode: if _with_bytecode {
                        bytecode_contracts[name].to_owned()
                    } else {
                        "".to_string()
                    },
                },
            )
        })
        .collect::<IndexMap<_, _>>();

    Ok(CompiledModule {
        src_ast,
        lowered_ast,
        contracts,
    })
}
