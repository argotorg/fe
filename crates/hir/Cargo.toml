[package]
name = "fe-hir"
version = "0.20.0-alpha"
authors = ["The Fe Developers <snakecharmers@ethereum.org>"]
edition = "2021"
license = "Apache-2.0"
repository = "https://github.com/ethereum/fe"
description = "Provides HIR definition and lowering for Fe lang"

[dependencies]
tracing = "0.1"
# We may need to fix this to a specific version, 
# but I want to keep up with the latest version until the new Fe implemeentation is merged into the master.
salsa = { git = "https://github.com/salsa-rs/salsa", package = "salsa-2022" }
derive_more = "0.99"
cranelift-entity = "0.91"
num-bigint = "0.4.3"
num-traits = "0.2.15"
camino = "1.1.4"
rustc-hash = "1.1.0"

parser = { path = "../parser2", package = "fe-parser2" }
common = { path = "../common2", package = "fe-common2" }
