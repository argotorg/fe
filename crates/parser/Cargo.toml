[package]
name = "fe-parser"
version = "0.26.0"
edition = "2021"
license = "Apache-2.0"
repository = "https://github.com/ethereum/fe"
description = "Parser lib for Fe."

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
derive_more.workspace = true
lazy_static = { workspace = true }
logos = { workspace = true }
rowan = { workspace = true }
rustc-hash.workspace = true
smallvec.workspace = true
unwrap-infallible = { workspace = true }

[dev-dependencies]
dir-test.workspace = true
wasm-bindgen-test.workspace = true
test-utils.workspace = true

[target.'cfg(target_arch = "wasm32")'.dependencies]
wasm-bindgen = "0.2"
