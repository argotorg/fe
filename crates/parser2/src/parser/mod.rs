use fxhash::FxHashSet;

use crate::{ParseError, SyntaxKind, TextRange};

use self::token_stream::{BackTrackableTokenStream, SyntaxToken, TokenStream};

pub mod token_stream;

/// Parser to build a rowan syntax tree.
pub struct Parser<S: TokenStream> {
    /// Token stream to parse.
    stream: BackTrackableTokenStream<S>,

    builder: rowan::GreenNodeBuilder<'static>,
    scopes: Vec<Box<dyn ParsingScope>>,
    errors: Vec<ParseError>,

    /// The checkpoint where the scope/branch is wrapped up later by another
    /// scope/branch when the `wrap_scope_with` method.
    check_point: Option<(rowan::Checkpoint, usize)>,

    current_pos: rowan::TextSize,
}

impl<S: TokenStream> Parser<S> {
    /// Returns the current token of the parser.
    pub fn current_token(&mut self) -> Option<&S::Token> {
        self.stream.peek()
    }

    /// Enters the scope and set the `scope` to the current scope.
    /// If `is_checkpoint` is true, the current scope/branch is wrapped up by
    /// another scope/branch later when the [`wrap_scope_with`] method is
    /// called.
    pub fn enter(&mut self, scope: Box<dyn ParsingScope>, is_checkpoint: bool) {
        if is_checkpoint {
            self.check_point = Some((self.builder.checkpoint(), self.scopes.len()));
        }

        self.builder.start_node(scope.syntax_kind().into());
        self.scopes.push(scope);
    }

    /// Enters the errors scope and add the `msg` to the error list.
    pub fn enter_with_error(&mut self, msg: &str) {
        let start = self.current_pos;
        let end = if let Some(current_token) = self.current_token() {
            start + current_token.text_size()
        } else {
            start
        };
        let range = TextRange::new(start, end);

        self.errors.push(ParseError {
            range,
            msg: msg.to_string(),
        });
        self.scopes.push(Box::new(ErrorScope()));
    }

    /// Leaves the current scope/branch.
    pub fn leave(&mut self) {
        self.scopes.pop();
        self.builder.finish_node();
    }

    /// Wrap up the marked scope/branch with another scope/branch and set the
    /// `scope` to the current scope.
    pub fn wrap_scope_with(&mut self, scope: Box<dyn ParsingScope>) {
        debug_assert!(self.check_point.is_some(), "No checkpoint");
        let check_point = self.check_point.take().unwrap();
        let syntax_kind = scope.syntax_kind();

        self.scopes.truncate(check_point.1);
        self.scopes.push(scope);

        self.builder
            .start_node_at(check_point.0, syntax_kind.into());
    }

    /// Bumps the current token and adds it to the current branch.
    pub fn bump(&mut self) {
        let tok = self.stream.next().unwrap();
        self.current_pos += rowan::TextSize::of(tok.text());
        self.builder.token(tok.syntax_kind().into(), tok.text());
    }

    /// Bumps consecutive trivia tokens.
    pub fn bump_trivias(&mut self) {
        while let Some(tok) = self.current_token() {
            let kind = tok.syntax_kind();
            if kind.is_trivia() {
                self.bump();
            } else {
                break;
            }
        }
    }

    /// Bump consecutive newlines.
    pub fn bump_newlines(&mut self) {
        while let Some(tok) = self.current_token() {
            if tok.syntax_kind() == SyntaxKind::Newline {
                self.bump();
            } else {
                break;
            }
        }
    }

    /// Proceeds the parser to the recovery token of the current scope. Then
    /// leave the current branch/scope.
    pub fn recovery(&mut self) {
        let mut scope_index = self.scopes.len() - 1;
        // Finds the nearest scope that has its own recovery set.
        loop {
            if self.scopes[scope_index].recovery_method() != &RecoveryMethod::Inheritance
                || scope_index == 0
            {
                break;
            } else {
                scope_index -= 1;
            }
        }

        while let Some(tok) = self.stream.peek() {
            let syntax_kind = tok.syntax_kind();
            if self.scopes[scope_index]
                .recovery_method()
                .contains(syntax_kind)
            {
                break;
            } else {
                self.bump();
            }
        }

        self.leave();
    }
}

/// The current scope of parsing.
pub trait ParsingScope {
    /// Returns the recovery method of the current scope.
    fn recovery_method(&self) -> &RecoveryMethod;

    fn syntax_kind(&self) -> SyntaxKind;
}

pub struct ErrorScope();

impl ParsingScope for ErrorScope {
    fn recovery_method(&self) -> &RecoveryMethod {
        &RecoveryMethod::Inheritance
    }

    fn syntax_kind(&self) -> SyntaxKind {
        SyntaxKind::Error
    }
}

/// Represents the recovery method of the current scope.
#[derive(PartialEq, Eq)]
pub enum RecoveryMethod {
    /// Uses the recovery method of the parent scope.
    Inheritance,

    /// The scope has its own recovery set.
    RecoverySet(FxHashSet<SyntaxKind>),
}

impl RecoveryMethod {
    /// Returns `true` if the recovery set contains the given syntax kind.
    fn contains(&self, syntax_kind: SyntaxKind) -> bool {
        match self {
            RecoveryMethod::Inheritance => false,
            RecoveryMethod::RecoverySet(set) => set.contains(&syntax_kind),
        }
    }
}

trait TextSize {
    fn text_size(&self) -> rowan::TextSize;
}

impl<T> TextSize for T
where
    T: SyntaxToken,
{
    fn text_size(&self) -> rowan::TextSize {
        rowan::TextSize::of(self.text())
    }
}
