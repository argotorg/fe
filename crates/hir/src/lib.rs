use common::{InputDb, Upcast};

pub mod hir_def;
pub mod input;
pub mod lower;
pub mod span;

#[salsa::jar(db = HirDb)]
pub struct Jar(
    // Tracked Hir items.
    hir_def::Mod,
    hir_def::Fn,
    hir_def::ExternFn,
    hir_def::Struct,
    hir_def::Contract,
    hir_def::Enum,
    hir_def::TypeAlias,
    hir_def::Impl,
    hir_def::Trait,
    hir_def::ImplTrait,
    hir_def::Const,
    hir_def::Use,
    // Interned structs.
    hir_def::Body,
    hir_def::IdentId,
    hir_def::IntegerId,
    hir_def::StringId,
    hir_def::PathId,
    hir_def::FnParamListId,
    hir_def::AttrListId,
    hir_def::WhereClauseId,
    hir_def::GenericArgListId,
    hir_def::GenericParamListId,
    hir_def::RecordFieldListId,
    hir_def::EnumVariantListId,
    hir_def::ImplItemListId,
    hir_def::TypeId,
    hir_def::UseTreeId,
    hir_def::ingot_module_tree,
    hir_def::module_item_tree,
);

pub trait HirDb: salsa::DbWithJar<Jar> + InputDb + Upcast<dyn InputDb> {}
impl<DB> HirDb for DB where DB: ?Sized + salsa::DbWithJar<Jar> + InputDb + Upcast<dyn InputDb> {}
