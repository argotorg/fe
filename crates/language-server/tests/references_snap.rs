use common::InputDb;
use dir_test::{dir_test, Fixture};
use driver::DriverDataBase;
use fe_semantic_query::SemanticIndex;
use hir::{lower::map_file_to_mod, span::{DynLazySpan, LazySpan}, SpannedHirDb};
use parser::SyntaxNode;
use test_utils::snap_test;
use url::Url;

// Collect cursor positions: identifiers, path/type path segments, field accessors
fn collect_positions(root: &SyntaxNode) -> Vec<parser::TextSize> {
    use parser::{ast, ast::prelude::AstNode, SyntaxKind};
    fn walk(node: &SyntaxNode, out: &mut Vec<parser::TextSize>) {
        match node.kind() {
            SyntaxKind::Ident => out.push(node.text_range().start()),
            SyntaxKind::Path => {
                if let Some(path) = ast::Path::cast(node.clone()) {
                    for seg in path.segments() {
                        if let Some(id) = seg.ident() { out.push(id.text_range().start()); }
                    }
                }
            }
            SyntaxKind::PathType => {
                if let Some(pt) = ast::PathType::cast(node.clone()) {
                    if let Some(path) = pt.path() {
                        for seg in path.segments() {
                            if let Some(id) = seg.ident() { out.push(id.text_range().start()); }
                        }
                    }
                }
            }
            SyntaxKind::FieldExpr => {
                if let Some(fe) = ast::FieldExpr::cast(node.clone()) {
                    if let Some(tok) = fe.field_name() { out.push(tok.text_range().start()); }
                }
            }
            _ => {}
        }
        for ch in node.children() { walk(&ch, out); }
    }
    let mut v = Vec::new();
    walk(root, &mut v);
    v.sort(); v.dedup(); v
}

fn line_col_from_cursor(cursor: parser::TextSize, s: &str) -> (usize, usize) {
    let mut line=0usize; let mut col=0usize;
    for (i, ch) in s.chars().enumerate() {
        if i == Into::<usize>::into(cursor) { return (line, col); }
        if ch == '\n' { line+=1; col=0; } else { col+=1; }
    }
    (line, col)
}

fn format_snapshot(content: &str, lines: &[String]) -> String {
    let header = content.lines().enumerate().map(|(i,l)| format!("{i:?}: {l}")).collect::<Vec<_>>().join("\n");
    let body = lines.join("\n");
    format!("{header}\n---\n{body}")
}

fn to_lsp_location_from_span(db: &dyn InputDb, span: common::diagnostics::Span) -> Option<async_lsp::lsp_types::Location> {
    let url = span.file.url(db)?;
    let text = span.file.text(db);
    let starts: Vec<usize> = text.lines().scan(0, |st, ln| { let o=*st; *st+=ln.len()+1; Some(o)}).collect();
    let idx = |off: parser::TextSize| starts.binary_search(&Into::<usize>::into(off)).unwrap_or_else(|n| n.saturating_sub(1));
    let sl = idx(span.range.start()); let el = idx(span.range.end());
    let sc: usize = Into::<usize>::into(span.range.start()) - starts[sl];
    let ec: usize = Into::<usize>::into(span.range.end()) - starts[el];
    Some(async_lsp::lsp_types::Location{ uri:url, range: async_lsp::lsp_types::Range{
        start: async_lsp::lsp_types::Position::new(sl as u32, sc as u32), end: async_lsp::lsp_types::Position::new(el as u32, ec as u32)
    }})
}

fn pretty_enclosing(db: &dyn SpannedHirDb, top_mod: hir::hir_def::TopLevelMod, off: parser::TextSize) -> Option<String> {
    let items = top_mod.scope_graph(db).items_dfs(db);
    let mut best: Option<(hir::hir_def::ItemKind, u32)> = None;
    for it in items {
        let lazy = DynLazySpan::from(it.span());
        let Some(sp) = lazy.resolve(db) else { continue };
        if sp.range.contains(off) {
            let w: u32 = (sp.range.end() - sp.range.start()).into();
            match best { None => best=Some((it,w)), Some((_,bw)) if w< bw => best=Some((it,w)), _=>{} }
        }
    }
    best.and_then(|(it,_)| hir::hir_def::scope_graph::ScopeId::from_item(it).pretty_path(db))
}

#[dir_test(dir: "$CARGO_MANIFEST_DIR/test_files", glob: "goto_*.fe")]
fn refs_snapshot_for_goto_fixtures(fx: Fixture<&str>) {
    let mut db = DriverDataBase::default();
    let file = db.workspace().touch(&mut db, Url::from_file_path(fx.path()).unwrap(), Some(fx.content().to_string()));
    let top = map_file_to_mod(&db, file);
    let green = hir::lower::parse_file_impl(&db, top);
    let root = SyntaxNode::new_root(green);
    let positions = collect_positions(&root);

    let mut lines = Vec::new();
    for cur in positions {
        let refs = SemanticIndex::find_references_at_cursor(&db, &db, top, cur);
        if refs.is_empty() { continue; }
        // Group refs by file basename with optional enclosing symbol
        use std::collections::{BTreeMap, BTreeSet};
        let mut grouped: BTreeMap<String, BTreeSet<String>> = BTreeMap::new();
        for r in refs {
            if let Some(sp) = r.span.resolve(&db) {
                if let Some(loc) = to_lsp_location_from_span(&db, sp.clone()) {
                    let path = loc.uri.path();
                    let fname = std::path::Path::new(path).file_name().and_then(|s| s.to_str()).unwrap_or(path);
                    let enc = pretty_enclosing(&db, top, sp.range.start());
                    let entry = match enc { Some(e) => format!("{} @ {}:{}", e, loc.range.start.line, loc.range.start.character),
                                            None => format!("{}:{}", loc.range.start.line, loc.range.start.character) };
                    grouped.entry(fname.to_string()).or_default().insert(entry);
                }
            }
        }
        let mut parts = Vec::new();
        for (f, set) in grouped.iter() { parts.push(format!("{}: {}", f, set.iter().cloned().collect::<Vec<_>>().join("; "))); }
        let (l,c) = line_col_from_cursor(cur, fx.content());
        lines.push(format!("cursor ({l}, {c}): {} refs -> {}", grouped.values().map(|s| s.len()).sum::<usize>(), parts.join(" | ")));
    }

    let snapshot = format_snapshot(fx.content(), &lines);
    // Write refs snapshot alongside goto snapshot per file
    let orig = std::path::Path::new(fx.path());
    let stem = orig.file_stem().and_then(|s| s.to_str()).unwrap_or("snapshot");
    let refs_name = format!("refs_{}.fe", stem);
    let refs_path = orig.with_file_name(refs_name);
    snap_test!(snapshot, refs_path.to_str().unwrap());
}

