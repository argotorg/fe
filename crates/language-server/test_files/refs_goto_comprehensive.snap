---
source: crates/language-server/tests/references_snap.rs
assertion_line: 128
expression: snapshot
---
0: // Test struct field access resolution
1: struct Container {
2:     pub value: u32
3: }
4: 
5: impl Container {
6:     pub fn get(self) -> u32 {
7:         self.value
8:     }
9: }
10: 
11: // Test enum variant resolution
12: enum Color {
13:     Red,
14:     Green { intensity: u32 },
15:     Blue(u32)
16: }
17: 
18: fn test_field_access() {
19:     let container = Container { value: 43 }
20: 
21:     // This should resolve to the field definition
22:     let val = container.value
23: 
24:     // This should resolve to the method definition
25:     let retrieved = container.get()
26: 
27:     // Test local variable references
28:     let copy_container = container  // should resolve to line 20
29:     let copy_val = val             // should resolve to line 23
30: }
31: 
32: fn test_enum_variants() {
33:     // These should resolve to the specific variants, not the enum
34:     let red = Color::Red
35:     let green = Color::Green { intensity: 50 }
36:     let blue = Color::Blue(100)
37: 
38:     // Test pattern matching field resolution
39:     match green {
40:         Color::Green { intensity } => {
41:             // 'intensity' here should resolve to the field in the enum variant
42:             let _val = intensity
43:         }
44:         _ => {}
45:     }
46: }
47: 
48: pub trait Inner {
49:     fn foo(self) -> i32
50: }
51: 
52: pub struct Wrapper<S: Inner> {
53:     pub inner: S,
54: }
55: 
56: impl<S: Inner> Wrapper<S> {
57:     pub fn foo(mut self, mut dog: i32) -> i32 {
58:         self.inner.foo() // collect_constraints_from_func_def
59:         dog
60:     }
61: }
62: 
---
cursor (7, 8): 2 refs -> goto_comprehensive.fe: 6:15; 7:8
cursor (7, 13): 1 refs -> goto_comprehensive.fe: goto_comprehensive::Container @ 2:8
cursor (19, 8): 4 refs -> goto_comprehensive.fe: goto_comprehensive::test_field_access::{fn_body} @ 19:8; goto_comprehensive::test_field_access::{fn_body} @ 22:14; goto_comprehensive::test_field_access::{fn_body} @ 25:20; goto_comprehensive::test_field_access::{fn_body} @ 28:25
cursor (22, 8): 2 refs -> goto_comprehensive.fe: goto_comprehensive::test_field_access::{fn_body} @ 22:8; goto_comprehensive::test_field_access::{fn_body} @ 29:19
cursor (22, 14): 4 refs -> goto_comprehensive.fe: goto_comprehensive::test_field_access::{fn_body} @ 19:8; goto_comprehensive::test_field_access::{fn_body} @ 22:14; goto_comprehensive::test_field_access::{fn_body} @ 25:20; goto_comprehensive::test_field_access::{fn_body} @ 28:25
cursor (22, 24): 1 refs -> goto_comprehensive.fe: goto_comprehensive::Container @ 2:8
cursor (25, 8): 1 refs -> goto_comprehensive.fe: goto_comprehensive::test_field_access::{fn_body} @ 25:8
cursor (25, 20): 4 refs -> goto_comprehensive.fe: goto_comprehensive::test_field_access::{fn_body} @ 19:8; goto_comprehensive::test_field_access::{fn_body} @ 22:14; goto_comprehensive::test_field_access::{fn_body} @ 25:20; goto_comprehensive::test_field_access::{fn_body} @ 28:25
cursor (28, 8): 1 refs -> goto_comprehensive.fe: goto_comprehensive::test_field_access::{fn_body} @ 28:8
cursor (28, 25): 4 refs -> goto_comprehensive.fe: goto_comprehensive::test_field_access::{fn_body} @ 19:8; goto_comprehensive::test_field_access::{fn_body} @ 22:14; goto_comprehensive::test_field_access::{fn_body} @ 25:20; goto_comprehensive::test_field_access::{fn_body} @ 28:25
cursor (29, 8): 1 refs -> goto_comprehensive.fe: goto_comprehensive::test_field_access::{fn_body} @ 29:8
cursor (29, 19): 2 refs -> goto_comprehensive.fe: goto_comprehensive::test_field_access::{fn_body} @ 22:8; goto_comprehensive::test_field_access::{fn_body} @ 29:19
cursor (34, 8): 1 refs -> goto_comprehensive.fe: goto_comprehensive::test_enum_variants::{fn_body} @ 34:8
cursor (34, 14): 1 refs -> goto_comprehensive.fe: goto_comprehensive::Color @ 12:5
cursor (34, 21): 2 refs -> goto_comprehensive.fe: goto_comprehensive::Color @ 13:4; goto_comprehensive::test_enum_variants::{fn_body} @ 34:21
cursor (35, 8): 2 refs -> goto_comprehensive.fe: goto_comprehensive::test_enum_variants::{fn_body} @ 35:8; goto_comprehensive::test_enum_variants::{fn_body} @ 39:10
cursor (36, 8): 1 refs -> goto_comprehensive.fe: goto_comprehensive::test_enum_variants::{fn_body} @ 36:8
cursor (36, 15): 1 refs -> goto_comprehensive.fe: goto_comprehensive::Color @ 12:5
cursor (36, 22): 2 refs -> goto_comprehensive.fe: goto_comprehensive::Color @ 15:4; goto_comprehensive::test_enum_variants::{fn_body} @ 36:22
cursor (39, 10): 2 refs -> goto_comprehensive.fe: goto_comprehensive::test_enum_variants::{fn_body} @ 35:8; goto_comprehensive::test_enum_variants::{fn_body} @ 39:10
cursor (40, 23): 2 refs -> goto_comprehensive.fe: goto_comprehensive::test_enum_variants::{fn_body} @ 40:23; goto_comprehensive::test_enum_variants::{fn_body} @ 42:23
cursor (42, 16): 1 refs -> goto_comprehensive.fe: goto_comprehensive::test_enum_variants::{fn_body} @ 42:16
cursor (42, 23): 2 refs -> goto_comprehensive.fe: goto_comprehensive::test_enum_variants::{fn_body} @ 40:23; goto_comprehensive::test_enum_variants::{fn_body} @ 42:23
cursor (58, 8): 2 refs -> goto_comprehensive.fe: 57:19; 58:8
cursor (58, 13): 1 refs -> goto_comprehensive.fe: goto_comprehensive::Wrapper @ 53:8
cursor (59, 8): 2 refs -> goto_comprehensive.fe: 57:29; 59:8
